File: plmem.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>
#include <string.h>

#define DEV_FILENAME "/dev/uio0"
#define MMAP_SIZE 0x1000
#define STRING_BLOCK_SIZE 0x7  // 7 bytes

static int parse_value(const char *val) {
  char *endptr;
  int result = (int)strtol(val, &endptr, 0);
  if (endptr[0] != '\0') return -1;
  return result;
}

// bit format
static int read_bit(const void *addr, int bit) {
  unsigned int val = ((const unsigned int*)addr)[0];
  printf("%d\n", (val >> bit) & 1);
  return 0;
}

static int write_bit(void *addr, int bit, int val) {
  if (val) {
    // set bit
    ((unsigned int*)addr)[0] |= 1 << bit;
  } else {
    // unset bit
    ((unsigned int*)addr)[0] &= ~(1 << bit);
  }
  return 0;
}

// byte format
static int read_byte(const void *addr) {
  printf("0x%02x\n", ((const unsigned char*)addr)[0]);
  return 0;
}

static int write_byte(void *addr, int val) {
  *((unsigned char*)addr) = (unsigned char)val;
  return 0;
}

// int16 format
static int read_int16(const void *addr) {
  printf("0x%08x\n", ((const uint16_t*)addr)[0]);
  return 0;
}

static int write_int16(void *addr, int val) {
  *((uint16_t*)addr) = (uint16_t)val;
  return 0;
}

// int32 format
static int read_int32(const void *addr) {
  printf("0x%08x\n", ((const unsigned int*)addr)[0]);
  return 0;
}

static int write_int32(void *addr, int val) {
  *((unsigned int*)addr) = (unsigned int)val;
  return 0;
}

// string format
static int read_string(const void *addr, int length) {
  int i;
  const unsigned char *addr2 = (const unsigned char *)addr;
  char *buf = (char *)malloc(2*length + 1);
  buf[length] = '\0';
  for (i = 0; i < length; ++i) {
    sprintf(buf + 2*i, "%02x", addr2[i]);
  }
  printf("%s\n", buf);
  free(buf);
  return 0;
}

static int write_string(void *addr, const char* str) {
  int i, val, result, ret;
  unsigned char *buf;
  int buf_size;
  int length = strlen(str);
  if (!length) return 0;  // success
  if (length % 2) {
    printf("Invalid string\n");
    return -1;
  }
  if ((str[0] == '0') && (str[1] == 'x')) {
    // skip 0x
    str += 2;
    length -= 2;
  }
  buf_size = length/2;
  buf = (unsigned char*)malloc(buf_size);
  result = 0;
  for (i = 0; i < length; i += 2) {
    ret = sscanf(str + i, "%2x", &val);
    if (ret == 0) {
      printf("Invalid string value at index %d: %2s\n", i, str + i);
      result = -1;
      break;
    }
    buf[i/2] = val;
  }
  if (!result) { // copy on success
#ifdef STRING_BLOCKS
    // copy string by blocks
    unsigned char *addr_ptr = (unsigned char*)addr;
    unsigned char *buf_ptr = (unsigned char*)buf;
    while (buf_size > 0) {
      memcpy(addr_ptr,
             buf_ptr,
             buf_size > STRING_BLOCK_SIZE ? STRING_BLOCK_SIZE : buf_size);
      addr_ptr += STRING_BLOCK_SIZE;
      buf_ptr += STRING_BLOCK_SIZE;
      buf_size -= STRING_BLOCK_SIZE;
    }
#else
    memcpy(addr, buf, buf_size);
#endif  // STRING_BLOCKS
  }
  free((void *)buf);
  return result;
}

// binary format
static int read_binary(const void *addr, int offset, int length) {
  const int block_size = sizeof(unsigned char) * 8;  // 8-bit blocks
  int i, idx, bit;
  unsigned char val;
  char *buf = (char *)malloc(length + 1);
  buf[length] = '\0';
  int block_offset = length%block_size;
  for (i = 0; i < length; ++i) {
    idx = (offset + i) / block_size;
    if (block_offset == 0) block_offset = block_size;
      bit = block_offset + offset - (i + 1 - idx * block_size);
    val = ((const unsigned char*)addr)[idx];
    buf[i] = ((val >> bit) & 1) ? '1' : '0';
  }
  printf("%s\n", buf);
  free((void *)buf);
  return 0;
}

static int write_binary(void *addr, int offset, const char* str) {
  const int block_size = sizeof(unsigned char) * 8;  // 8-bit blocks
  int i, idx, bit;
  int length = strlen(str);
  if (!length) return 0;  // success
  // check bit string
  for (i = 0; i < length; ++i) {
    if ((str[i] != '0') && (str[i] != '1')) return -1;
  }
  for (i = 0; i < length; ++i) {
    idx = (offset + i) / block_size;
    int block_offset = length%block_size;
    if (block_offset == 0) block_offset = block_size;
      bit = block_offset + offset - (i + 1 - idx * block_size);
    if (str[i] != '0') {
      // set bit
      ((unsigned char*)addr)[idx] |= 1 << bit;
    } else {
      // unset bit
      ((unsigned char*)addr)[idx] &= ~(1 << bit);
    }
  }
  return 0;
}

static void usage(const char *progname) {
  printf("Usage:\n"
         "  plmem get address { bit num | byte | int16 | int32 | string length | binary offset length }\n"
         "  plmem set address { bit num | byte | int16 | int32 | string | binary offset } value\n"
         "\n"
         "Version 0.6 21.12.2014\n");
}

int main (int argc, char* argv[]) {
  int fd;
  void *map_addr;
  volatile unsigned char *mapped;
  void *target_addr;
  int val;
  int result;

  if (argc < 4) {
    printf("Too few arguments\n");
    usage(argv[0]);
    return 1;
  }

  enum {CMD_GET, CMD_SET} cmd;

  val = 0;
  if (!strcmp(argv[1], "get")) {
    cmd = CMD_GET;
  } else if (!strcmp(argv[1], "set")) {
    cmd = CMD_SET;
    if (argc < 5) {
      printf("Missing value for set command\n");
      usage(argv[0]);
      return 1;
    }
    val = parse_value(argv[4]);
  } else {
    printf("Invalid command: %s\n", argv[1]);
    usage(argv[0]);
    return 1;
  }

  int offset = parse_value(argv[2]);
  if (offset < 0) {
    printf("Invalid address: %s\n", argv[2]);
    return 2;
  }

  fd = open(DEV_FILENAME, O_RDWR);
  if (fd < 0) {
    perror("Failed to open devfile");
    return 1;
  }

  map_addr = mmap(NULL, MMAP_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
  if (map_addr == MAP_FAILED) {
    perror("Failed to mmap");
    close(fd);
    return 1;
  }

  mapped = map_addr;
  target_addr = (void *)(mapped + offset);

  result = -1;
  if (!strcmp(argv[3], "bit")) {
    int bit_num = (argc > 4) ? parse_value(argv[4]) : 0;
    if ((bit_num >= 0) && (bit_num < 32)) {
      if (cmd == CMD_GET) {
        result = read_bit(target_addr, bit_num);
      } else {  // CMD_SET
        val = (argc > 5) ? parse_value(argv[5]) : 0;
        if ((val == 0) || (val == 1)) {
          result = write_bit(target_addr, bit_num, val);
        } else {
          printf("Invalid bit value\n");
        }
      }
    } else if (bit_num < 0) {
      printf("Invalid bit number: %s\n", argv[4]);
    } else {
      printf("Bit number %d not in range [0; 31]\n", bit_num);
    }
  } else if (!strcmp(argv[3], "byte")) {
    if (cmd == CMD_GET) {
      result = read_byte(target_addr);
    } else {  // CMD_SET
      result = write_byte(target_addr, val);
    }
  } else if (!strcmp(argv[3], "int16")) {
    if (cmd == CMD_GET) {
      result = read_int16(target_addr);
    } else {  // CMD_SET
      result = write_int16(target_addr, val);
    }
  } else if (!strcmp(argv[3], "int32")) {
    if (cmd == CMD_GET) {
      result = read_int32(target_addr);
    } else {  // CMD_SET
      result = write_int32(target_addr, val);
    }
  } else if (!strcmp(argv[3], "string")) {
    if (cmd == CMD_GET) {
      int length = (argc > 4) ? parse_value(argv[4]) : 0;
      if (length > 0) {
        result = read_string(target_addr, length);
      } else {
        printf("Wrong string length value\n");
      }
    } else {  // CMD_SET
      result = write_string(target_addr, argv[4]);
    }
  } else if (!strcmp(argv[3], "binary")) {
    int bit_offset = (argc > 4) ? parse_value(argv[4]) : 0;
    if (cmd == CMD_GET) {
      int length = (argc > 5) ? parse_value(argv[5]) : 0;
      if (length > 0) {
        result = read_binary(target_addr, bit_offset, length);
      } else {
        printf("Wrong binary length value\n");
      }
    } else {  // CMD_SET
      if (argc > 5) {
        result = write_binary(target_addr, bit_offset, argv[5]);
      } else {
        printf("Missing binary value\n");
      }
    }
  }

  munmap(map_addr, MMAP_SIZE);
  close(fd);
  return result;
}

File: plmem.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 20 10:23:28 2015

@author: drvmotor
TITLE: python3.3 equivlent for Xillubus operations
all pieces of orig. C-source are stripped

# include <stdio.h>
# include <stdlib.h>
# include <stdint.h>
# include <fcntl.h>
# include <unistd.h>
# include <sys/mman.h>
# include <string.h>

# define DEV_FILENAME "/dev/uio0"
# define MMAP_SIZE 0x1000
# define STRING_BLOCK_SIZE 0x7 // 7 bytes
"""
import sys
import mmap

if __name__ == '__main__':
    # Actual device
    DEV_FILENAME = "/dev/uio0"
else:
    # dummy device for unit test
    DEV_FILENAME = "dvc"

MMAP_SIZE = 0x1000
STRING_BLOCK_SIZE = 0x07
CMD_GET = 0
CMD_SET = 1

value = 0
cmd = CMD_GET

def parse_value(strval, errstr):
 try:
    res = int(strval, 0)
 except ValueError:
    print(errstr)
    usage(sys.argv[0])
    sys.exit(1)
 return res


class Xillybus():
	# xmap=0

	def __init__(self, dev_fn=''):
		fdev = open(dev_fn, 'r+b')
		self.xmap = mmap.mmap(
		    fdev.fileno(), MMAP_SIZE, mmap.MAP_SHARED, mmap.PROT_READ | mmap.PROT_WRITE, 0)
	def read_bit(self, ptr, bit):
		ba = bit >> 3;
		self.xmap.seek(ptr + ba)
		bb = (self.xmap.read_byte() >> (bit & 7))
		print("{0}\n".format(bb & 1))
		return 0

	def write_bit(self, ptr, bit, val):
		ba = bit >> 3
		bit &= 7
		self.xmap.seek(ptr + ba)
		bb = self.xmap.read_byte()
		if val == 0:
			bb &= ~(1 << bit)
		else:
			bb &= 1 << bit
		self.xmap.seek(ptr + ba)
		self.xmap.write_byte((bb))
		return 0

	def read_byte(self,ptr):
		self.xmap.seek(ptr)
		bbb=self.xmap.read_byte()
		print("{0}\n".format(hex(bbb)))
		return 0

	def write_byte(self,ptr,val):
		self.xmap.seek(ptr)
		self.xmap.write_byte((val & 255))
		return 0

	def read_int16(self,ptr):
		self.xmap.seek(ptr)
		#bbb = ord(self.xmap.read_byte())
		bbb = self.xmap.read_byte() & 255
		#bbc = ord(self.xmap.read_byte())
		bbc = self.xmap.read_byte() & 255
		bbd = bbb+(bbc*256)
		print("{0}\n".format(hex(bbd)))
		return 0

	def write_int16(self,ptr,val):
		self.xmap.seek(ptr)
		self.xmap.write_byte((val & 255))
		self.xmap.write_byte((val>>8))
		return 0

	def read_int32(self,ptr):
		self.xmap.seek(ptr)
		bbb= self.xmap.read_byte()
		bbb+= self.xmap.read_byte()<<8
		bbb+= self.xmap.read_byte()<<16
		bbb+= self.xmap.read_byte()<<24
		print("{0}\n".format(hex(bbb)))
		return 0
   
	def write_int32(self,ptr,val):
		self.xmap.seek(ptr)
		self.xmap.write_byte(val & 255)
		self.xmap.write_byte((val>>8) & 255)
		self.xmap.write_byte((val>>16) & 255)
		self.xmap.write_byte((val>>24)& 255)
		return 0

	def read_string(self,ptr,length):
		self.xmap.seek(ptr)
		s=''
		while len(s)<length*2:
			d1 = hex(self.xmap.read_byte())
			d = d1[2:4]
			if len(d) < 2 :
				d= '0'+ d
			s += d
		print( "{0}\n".format(s))
		return 0

	def h2bin(self,s,j):
		x=s[j]
		if (x >= '0') and (x <= '9'):
			return ord(x)-ord('0')
		elif (x>='A') and (x <= 'F'):
			return ord(x)-ord('A')+10
		else:
			return -1
			
	def write_string(self,ptr,sss):
		if (len(sss) & 1) != 0:
			print( "Invalid string\n")
			sys.exit(1)
		SSS=sss.upper()
		ddd=[]
		for j in range(0,len(SSS),2):
			dgh=self.h2bin(SSS,j)
			dgl=self.h2bin(SSS,j+1)
			if (dgh < 0) or (dgl < 0):
				print( "Illegal string value at index {0}: {1}\n".format(j,sss))
			ddd += [dgh*16 + dgl]
		self.xmap.seek(ptr)
		for x in ddd:
			self.xmap.write_byte((x & 255))
		return 0
		
	def read_bin(self,ptr,offset,length):
		self.xmap.seek(ptr+offset//8)
		binbuf=''
		cnt=0
		offset &= 7
		byteval=0
		while length >0:
			length-=1
			if cnt == 0:
				byteval=self.xmap.read_byte() >> offset
				cnt = 8-offset
				offset = 0
			cnt-=1
			sym=chr((byteval & 1)+ ord('0'))
			byteval >>=1
			binbuf +=sym
		print( "{0}\n".format(binbuf))
		return 0 

	def write_bin(self,ptr,offset,sss):
		if len(sss)==0:
			return 0
		for x in sss:#validation
			if (x != '0') and (x != '1'):
				return -1
		# position
		begin = ptr+offset//8 
		offset &= 7
		self.xmap.seek(begin)
		begval=self.xmap.read_byte()
		self.xmap.seek(begin)#restore pos
		for sym in sss:
			msk = ~(1 << offset)
			bitv = ord(sym)-ord('0')
			begval &= msk
			begval |= bitv << offset
			offset += 1
			if offset > 7:
				# store byte
				self.xmap.write_byte(begval)
				begval = 0
				offset = 0
		if offset != 0:
			self.xmap.write_byte(begval)
		return 0

def usage(s=""):
	print("Usage:\n")
	print(s," get address { bit num | byte | int16 | int32 | string length | binary offset length }\n")
	print(s," set address { bit num | byte | int16 | int32 | string | binary offset } value\n")

def get_func(addr=0,xil=0):
	pass
def set_func(addr=0,xil=0):
	pass  
	    
def main():
#  """
# if (argc < 4) {
# printf("Too few arguments\n");
# usage(argv[0]);
# return 1;
# }
# """
	retcode =0
	if len(sys.argv) < 4:
	    print("Too few arguments\n")
	    usage(sys.argv[0])
	    sys.exit(1)
	# check addr
	address= parse_value(sys.argv[2],"Invalid address: {0}\n".format(sys.argv[2]))
	# memmap device
	xil=Xillybus(DEV_FILENAME)
	# check cmd
	if sys.argv[1]=="get":
		retcode = get_func(address,xil)
	elif sys.argv[1]=="set":
		if len(sys.argv) < 5:
			print( "Missing value for set command\n")
			usage(sys.argv[0])
			sys.exit(1)
		retcode = set_func(address,xil)
	else: #just error
		print("Invalid command: {0}\n".format(sys.argv[1]))
		usage(sys.argv[0])
		sys.exit(1)
	
	return retcode
	
# read a number from a source and validate it for max and min values 
def get_anum(s,nmin,nmax):
	num=parse_value(s,"Invalid number: {0}\n".format(s))
	if (num < nmin) or (num > nmax):
		print("Number {0} not in range [{1}..{2}]\n".format(num,nmin,nmax))
		sys.exit(1)
	return num
	
def get_bitnum(s):
	num=parse_value(s,"Invalid bit number: {0}\n".format(s))
	if (num < 0) or (num > 31):
		print("Bit number {0} not in range [0; 31]\n".format(num))
		sys.exit(1)
	return num
    
	
def get_func(addr=0,xil=0):
	if sys.argv[3]=='bit':
		bitn = get_bitnum(sys.argv[4])
		return xil.read_bit(addr,bitn)
	if sys.argv[3]=='byte':
		return xil.read_byte(addr)
	if sys.argv[3]=='int32':
		return xil.read_int32(addr)
	if sys.argv[3]=='int16':
		return xil.read_int16(addr)
	if sys.argv[3]=='string':
		lng = parse_value(sys.argv[4],"Invalid string length: {0}".format(sys.argv[4]))
		return xil.read_string(addr,lng)
	if sys.argv[3]=='binary':
		offs=get_bitnum(sys.argv[4])
		lngof=get_anum(sys.argv[5],0,0xFFFFFFFF)
		return xil.read_bin(addr,offs,lngof)
    # empty action
	print("Warning: nothing to be done\n")
	return 0
	
def set_func(addr=0,xil=0):
    # check value
	if sys.argv[3]=='bit':
	# read number
		bitn = get_bitnum(sys.argv[4])
		val= ord(sys.argv[5])-ord('0')
		return xil.write_bit(addr,bitn,val)
	if sys.argv[3]=='byte':
		val = get_anum(sys.argv[4],0,255)
		return xil.write_byte(addr,val)
	if sys.argv[3]=='int32':
		val = get_anum(sys.argv[4],0,0xFFFFFFFF)
		return xil.write_int32(addr,val)
	if sys.argv[3]=='int16':
		val = get_anum(sys.argv[4],0,0xFFFF)
		return xil.write_int16(addr,val)
	if sys.argv[3]=='string':
		return xil.write_string(addr,sys.argv[4])
	if sys.argv[3]=='binary':
		offs=get_bitnum(sys.argv[4])
		return xil.write_bin(addr,offs,sys.argv[5])
    # empty action
	print("Warning: nothing to be done\n")
	return 0
# start
if __name__ == '__main__':
	main()
